### PR TITLE
Metabase 43 QP middleware overhaul part 1: split pre+post middleware into separate fns

### PR DESCRIPTION
**Pull request by camsaul**
Step 1 of #19959. See #19959 for background and details.

This is a partial version of that PR. This PR splits all of our middleware that did both pre- and post- processing into separate pre-processing and post-processing middleware (there were only four or five middleware functions that needed updating). This is actually a little bit of a tricky change because this sort of 'around' middleware was previously able to maintain state inside the middleware itself, whereas now it needs to be saved as part of the pre-processed query, which I've done using namespaced keywords. This ultimately makes things more transparent and easier to test, since pre-processing and post-processing can be tested independently of one another.

I've added some utility functions that wrap middleware written in the new 43+ style to be used as <=42 middleware. I wrote the new middleware as 43+ middleware and used these functions for compatibility; the intention is that in a follow-on PR we can simple remove the wrapper functions and use the 43+ versions directly.

Next steps:

- split pre-processing, post-processing, compilation, and "execution" (permissions check/caching/etc.) middleware into separate groups 
- rework `query->preprocessed` to just use pre-processing middleware; `query->native` to use preprocessing middleware + `mbql->native`; remove `metabase.query-processor.reducible/quit-early-exception` and `metabase.query-processor.context/preprocessedf`, and `nativef`
- rework remaining pre & post middleware to use the new 43 patterns described in #19855

### CODE REVIEW

File: enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj [hunk 1]
```diff
@@ -301,38 +303,54 @@
         (_ :guard (every-pred map? :source-table))
         (assoc &match ::gtap? true)))))
 
+(defn- expected-cols [query]
+  (binding [*current-user-permissions-set* (atom #{"/"})]
+    ((requiring-resolve 'metabase.query-processor/query->expected-cols) query)))
```
Comment (dpsutton) [inline enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj:308]: Just a note, should we use a notion of the `classloader/require` here versus "plain" require?

File: src/metabase/query_processor.clj [hunk 2]
```diff
@@ -74,58 +74,63 @@
...
-   #'cache/maybe-return-cached-results
-   #'optimize-temporal-filters/optimize-temporal-filters
-   #'validate-temporal-bucketing/validate-temporal-bucketing
-   #'auto-parse-filter-values/auto-parse-filter-values
-   #'wrap-value-literals/wrap-value-literals
-   #'annotate/add-column-info
-   #'perms/check-query-permissions
-   #'pre-alias-ags/pre-alias-aggregations
-   #'cumulative-ags/handle-cumulative-aggregations
+  [#'mbql-to-native/mbql->native                                      ; EXECUTE
```
Comment (dpsutton) [inline src/metabase/query_processor.clj:80]: I guess too late but I think I would have done something like

```clojure
(m.43/wrap #'limit/limit-result-rows-middleware #'limit/add-default-limit-middleware)
(m.43/wrap identity #'annotate/add-column-info)
```

then when we undo this the middleware's don't change, and we're sure we are using the same vars in the new middleware stacks that will emerge. And the comments cannot drift or be incorrect.

**Review by dpsutton — APPROVED** [review]

File: enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj (hunk 1 above)
Comment (camsaul) [reply enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj:308]: Yeah good call

File: src/metabase/query_processor.clj (hunk 2 above)
Comment (camsaul) [reply src/metabase/query_processor.clj:80]: yeah that actually does look like a better idea in hindsight

**Pull request merged**